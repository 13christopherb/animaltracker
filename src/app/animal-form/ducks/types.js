//@flow

import {actionTypes} from "./action-types";

export type Animal = {
    id: string,
    name: string,
    weight: string,
    location: string,
    isGettingTubed: boolean,
    isGettingControlledMeds: boolean
}

export type Field = {
    name: string,
    value: string | boolean
}

export type State = {
    +animals: Array<{
        +id: string,
        +name: string,
        +weight: string,
        +location: string,
        +isGettingTubed: boolean,
        +isGettingControlledMeds: boolean
    }>,
    +fields: {
        +name: string,
        +weight: string,
        +location: string,
        +isGettingTubed: boolean,
        +isGettingControlledMeds: boolean
    }
};

export type ChangeFieldAction = {
    type: 'CHANGE_FIELD_VALUE',
    field: {name: string, value: string | boolean}
}

export type AddAnimalSuccessAction = {
    type: 'ADD_ANIMAL_SUCCESS',
    animal: {
        id: string,
        name: string,
        weight: string,
        location: string,
        isGettingTubed: boolean,
        isGettingControlledMeds: boolean}
}

export type AddAnimalFailureAction = {
    type: 'ADD_ANIMAL_FAILURE',
}

export type AddAnimalRequestAction = {
    type: 'ADD_ANIMAL_REQUEST',
    animal: {
        id: string,
        name: string,
        weight: string,
        location: string,
        isGettingTubed: boolean,
        isGettingControlledMeds: boolean}
}

export type Action = ChangeFieldAction | AddAnimalSuccessAction | AddAnimalFailureAction | AddAnimalRequestAction

export type Dispatch = (action: Action | ThunkAction | PromiseAction) => any;
export type GetState = () => State;
export type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;
export type PromiseAction = Promise<Action>;