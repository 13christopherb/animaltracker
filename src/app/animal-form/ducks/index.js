/** @module animal-form */
import animalFormReducer from './reducers';
export {actions as animalFormActions} from './actions';
export {types as animalFormTypes} from './types';
export default animalFormReducer;