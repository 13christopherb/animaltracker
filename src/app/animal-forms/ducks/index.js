/** @module animal-form */
export {types as animalFormTypes} from './types';