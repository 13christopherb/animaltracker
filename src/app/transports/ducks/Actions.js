import * as transportService from './Operations';
import { types} from "./Types";

export const actions = {
    addTransport,
    getTransports,
    deleteTransport
};

function addTransport(transport) {
    return dispatch => {
        dispatch(request(transport));
        return transportService.addTransport(transport)
            .then(res => {
                    dispatch(success(res))
                },
                error=>{
                    dispatch(failure(error))
                });
    };
    function request(transport) {
        return {type: types.ADD_TRANSPORT_REQUEST, transport}
    }

    function success(transport) {
        return {type: types.ADD_TRANSPORT_SUCCESS, transport}
    }

    function failure(error) {
        return {type: types.ADD_TRANSPORT_FAILURE, error}
    }
}

function getTransports() {
    return dispatch => {
        dispatch(request());
        return transportService.getAllTransports()
            .then(
                res => {
                    dispatch(success(res['transports']))
                },
                error => {
                    dispatch(failure(error))
                }
            );
    };
    function request() {
        return {type: types.GET_TRANSPORTS_REQUEST}
    }

    function success(transports) {
        return {type: types.GET_TRANSPORTS_SUCCESS, transports}
    }

    function failure(error) {
        return {type: types.GET_TRANSPORTS_FAILURE, error}
    }
}

function deleteTransport(transport) {
    return dispatch => {
        dispatch(request(transport));
        return transportService.deleteTransport(transport.id)
            .then(
                res => {
                    dispatch(success(transport))
                },
                error => {
                    dispatch(failure(error))
                }
            );
    };
    function request(transport) {
        return {type: types.DELETE_TRANSPORT_REQUEST, transport}
    }

    function success(transport) {
        return {type: types.DELETE_TRANSPORT_SUCCESS, transport}
    }

    function failure(error) {
        return {type: types.DELETE_TRANSPORT_FAILURE, error}
    }
}