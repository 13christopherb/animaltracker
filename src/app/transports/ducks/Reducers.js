import {types} from "./Types";
import _ from "underscore";

const initialState = {
    transports: []
};

export default function transportsReducer(state=initialState, action) {
    switch (action.type) {
        case types.ADD_TRANSPORT_SUCCESS:
            let transports = [...state['transports']];
            transports.push(action.transport);
            return {
                ...state,
                transports: transports
            };
        case types.GET_TRANSPORTS_SUCCESS:
            return {
                ...state,
                transports: action.transports
            };
        case types.DELETE_TRANSPORT_SUCCESS:
            return {
                ...state,
                transports: _.reject(state['transports'], (t) => {
                    return action.transport.id === t.id;
                })
            };
        default:
            return state
    }
}