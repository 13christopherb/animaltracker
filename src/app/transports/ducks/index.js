import transportsReducer from './Reducers';
export {actions as transportActions} from './Actions';
export {types as transportTypes} from './Types';
export default transportsReducer;