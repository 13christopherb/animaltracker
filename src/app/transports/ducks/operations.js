import {apiInstance} from "../../../utils/api";

export const getAllTransports = () =>
    apiInstance.get('/transports').then(res => res.data);

export const deleteTransport = (id) =>
    apiInstance.delete('/transport/'+id);