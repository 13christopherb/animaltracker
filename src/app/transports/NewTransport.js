import React, {Component} from 'react';
import {connect} from 'react-redux';
import DatePicker from 'react-datepicker';
import moment from 'moment';
import {actions} from "./ducks/actions";

import 'react-datepicker/dist/react-datepicker.css';

class NewTransport extends Component {

    constructor(props) {
        super(props);
        this.state = {
            value: '',
            departs: '',
            arrives: '',
            meetTime: this.props.currentTime
        };
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleDateTimeChange = this.handleDateTimeChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    componentDidMount() {
        //this.props.dispatch(actions.fetchAnimals());
    }


    handleInputChange(e) {
        const target = e.target;
        const value = target.value;
        const name = target.name;
        this.setState({
            [name]: value
        });
    }

    handleDateTimeChange(dateTime) {
        this.setState({
            meetTime: dateTime
        });
    }

    handleSubmit(e) {
        e.preventDefault();
        const transport = {
            departs: this.state.departs,
            arrives: this.state.arrives,
            meetTime: this.state.meetTime.format()
        };
        this.props.dispatch(actions.addTransport(transport));
    }

    render() {
        return (
            <div>
                <form onSubmit={this.handleSubmit}>
                    <div className="form-row">
                        <div className="col col-md-3">
                            <label>
                                Departs
                                <select name="departs" className="custom-select"
                                        value={this.state.departs} onChange={this.handleInputChange}>
                                    <option value=""></option>
                                    <option value="NRO">NRO</option>
                                    <option value="MBO">MBO</option>
                                    <option value="SLO">SLO</option>
                                </select>
                            </label>
                        </div>
                        <div className="col col-md-3">
                            <label>
                                Arrives
                                <select name="arrives" className="custom-select"
                                        value={this.state.arrives} onChange={this.handleInputChange}>
                                    <option value=""></option>
                                    <option value="NRO">NRO</option>
                                    <option value="MBO">MBO</option>
                                    <option value="SLO">SLO</option>
                                </select>
                            </label>
                        </div>
                        <div className="col col-md-3">
                            <label>
                                Meet Time
                                <DatePicker
                                    name="meetDate"
                                    selected={this.state.meetTime}
                                    onChange={this.handleDateTimeChange}
                                    showTimeSelect
                                    dateFormat="MM/DD/YYYY"
                                    placeholderText="Click to select a date"
                                    timeFormat='HH:mm'
                                    minDate={moment()}
                                    maxDate={moment().add(5, "days")}
                                    minTime={moment().hours(8).minutes(0)}
                                    maxTime={moment().hours(20).minutes(0)}
                                    autoComplete="off"
                                />
                            </label>
                        </div>
                        <button className="btn btn-primary" type="submit">Submit</button>
                    </div>
                </form>
            </div>
        );
    }
}

function mapStateToProps({}, ownProps) {
    return {
        meetTime: ownProps.currentTime
    }
}

const connectedNewTransport = connect(
    mapStateToProps,
)(NewTransport);

export {connectedNewTransport as NewTransport};