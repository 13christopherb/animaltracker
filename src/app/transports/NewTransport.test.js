import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import thunk from 'redux-thunk';
import moment from 'moment';
import {NewTransport} from "./NewTransport";
import {types} from "./ducks/types";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const initialState = {
    transports: []
};


const store = mockStore(initialState);

describe('<NewTransport />', () => {
    describe('render()', () => {
        let props = {
            currentTime: moment('2017-09-15 09:30:00')
        };
        test('should render the component', () => {
            const wrapper = shallow(<NewTransport {...props} store={store}/>);
            const component = wrapper.dive();
            expect(toJson(component)).toMatchSnapshot()
        })
    });

    describe('Creating new transportsReducer', () => {
        let props = {
            currentTime: moment('2017-09-15 09:30:00')
        };
        describe('user inputs are echoed', () => {
            test('departs select is echoed', () => {
                const wrapper = shallow(<NewTransport {...props} store={store}/>);
                const component = wrapper.dive();
                component.find('select[name="departs"]').simulate('change', {
                    target: {
                        value: 'new text',
                        type: 'test',
                        name: 'departs'
                    }
                });
                expect(component.find('select[name="departs"]').props().value).toEqual('new text');
            });
            test('arrives select is echoed', () => {
                const wrapper = shallow(<NewTransport {...props} store={store}/>);
                const component = wrapper.dive();
                component.find('select[name="arrives"]').simulate('change', {
                    target: {
                        value: 'new text',
                        type: 'test',
                        name: 'arrives'
                    }
                });
                expect(component.find('select[name="arrives"]').props().value).toEqual('new text');
            });
        });

        test('should prevent default browser submit event', () => {
            const wrapper = shallow(<NewTransport {...props} store={store}/>);
            const component = wrapper.dive();
            let prevented = false;
            component.find('form').simulate('submit', {
                preventDefault: () => {
                    prevented = true;
                }
            });
            expect(prevented).toBe(true);
        });

        test('should fire ADD_ANIMAL action', () => {
            const wrapper = shallow(<NewTransport {...props} store={store}/>);
            const component = wrapper.dive();
            const expectedActions = [
                {type: types.ADD_TRANSPORT_REQUEST}
            ];
            component.find('form').simulate('submit', {
                preventDefault: () => {
                    return false;
                }
            });
            expect(store.getActions()[0]['type']).toEqual(expectedActions[0]['type']);
        })
    });

});
