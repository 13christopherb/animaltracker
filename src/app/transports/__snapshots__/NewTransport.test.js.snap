// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewTransport /> render() should render the component 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <div
        className="col col-md-3"
      >
        <label>
          Departs
          <select
            className="custom-select"
            name="departs"
            onChange={[Function]}
            value=""
          >
            <option
              value=""
            />
            <option
              value="NRO"
            >
              NRO
            </option>
            <option
              value="MBO"
            >
              MBO
            </option>
            <option
              value="SLO"
            >
              SLO
            </option>
          </select>
        </label>
      </div>
      <div
        className="col col-md-3"
      >
        <label>
          Arrives
          <select
            className="custom-select"
            name="arrives"
            onChange={[Function]}
            value=""
          >
            <option
              value=""
            />
            <option
              value="NRO"
            >
              NRO
            </option>
            <option
              value="MBO"
            >
              MBO
            </option>
            <option
              value="SLO"
            >
              SLO
            </option>
          </select>
        </label>
      </div>
      <div
        className="col col-md-3"
      >
        <label>
          Meet Time
          <DatePicker
            allowSameDay={false}
            autoComplete="off"
            dateFormat="MM/DD/YYYY"
            dateFormatCalendar="MMMM YYYY"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            maxDate={"2018-10-31T00:18:27.243Z"}
            maxTime={"2018-10-26T03:00:27.244Z"}
            minDate={"2018-10-26T00:18:27.243Z"}
            minTime={"2018-10-25T15:00:27.244Z"}
            monthsShown={1}
            name="meetDate"
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            placeholderText="Click to select a date"
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            selected={"2017-09-15T16:30:00.000Z"}
            shouldCloseOnSelect={true}
            showTimeSelect={true}
            timeCaption="Time"
            timeFormat="HH:mm"
            timeIntervals={30}
            withPortal={false}
          />
        </label>
      </div>
      <button
        className="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
